zap/src/config.rs

```diff
@@ -261,27 +270,50 @@ impl<'src> Ty<'src> {
 			Self::Set(..) => (self.variants_size().unwrap().size(), None),
 
 			Self::Opt(ty) => {
-				let (_, ty_max) = ty.size(tydecls, recursed);
+				let (_, ty_max) = ty.size(recursed);
 
 				(1, ty_max.map(|ty_max| ty_max + 1))
 			}
 
-			Self::Ref(name, ..) => {
+			Self::Ref(name, tydecl) => {
 				if recursed.contains(name) {
 					// 0 is returned here because all valid recursive types are
 					// bounded and all bounded types have their own min size
 					(0, None)
 				} else {
 					recursed.insert(name);
 
-					let tydecl = tydecls.get(name).unwrap();
-
-					tydecl.size(tydecls, recursed)
+					tydecl.size(recursed)
 				}
 			}
 
-			Self::Enum(enum_ty) => enum_ty.size(tydecls, recursed),
-			Self::Struct(struct_ty) => struct_ty.size(tydecls, recursed),
+			Self::Enum(enum_ty) => enum_ty.size(recursed),
+			Self::Struct(struct_ty) => struct_ty.size(recursed),
+			Self::Or(or_tys, discriminant_numty) => {
+				let mut min = 0;
+				let mut max = Some(0usize);
+
+				for ty in or_tys {
+					let (ty_min, ty_max) = ty.size(recursed);
+
+					if ty_min < min {
+						min = ty_min;
+					}
+
+					if let Some(ty_max) = ty_max {
+						if let Some(current_max) = max {
+							max = Some(ty_max + current_max);
```

Comment:
Doesn't this needs to check if what max is being set to is larger than the current value of max?